File: src/main/java/org/dbshell/db/metadata/TableProvider.java
package org.dbshell.db.metadata;

import org.dbshell.commands.connections.dto.ConnectionInfo;
import org.dbshell.commands.connections.dto.ConnectionInfoUtil;
import org.dbshell.environment.EnvironmentVars;
import org.springframework.core.MethodParameter;
import org.springframework.shell.CompletionContext;
import org.springframework.shell.CompletionProposal;
import org.springframework.shell.standard.ValueProviderSupport;
import org.springframework.stereotype.Component;

import java.sql.DatabaseMetaData;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

@Component
public class TableProvider extends ValueProviderSupport {

    @Override
    public List<CompletionProposal> complete(MethodParameter parameter, CompletionContext completionContext, String[] hints) {
        String currentInput = completionContext.currentWord();
        List<CompletionProposal> proposals = new ArrayList<>();
        String[] types = new String[] {"TABLE"};

        try {
            DatabaseMetaData dbmd = ConnectionInfoUtil.getConnectionFromCurrentContextJndi().getConnection().getMetaData();
            String catalog = EnvironmentVars.getCurrentCatalog();
            String schema = EnvironmentVars.getCurrentSchema();

            return DatabaseMetadata.getTables(dbmd, catalog, schema, types)
                    .stream()
                    .map(t -> t.getTableName())
                    .filter(s -> s.contains(currentInput))
                    .map(CompletionProposal::new)
                    .collect(Collectors.toList());

        } catch(SQLException sqlEx) {
            sqlEx.getMessage();
            return proposals;
        }
    }
}

File: src/main/kotlin/org/dbshell/providers/TableProvider.kt
package org.dbshell.providers

import org.dbshell.shellmethods.dto.ConnectionInfoUtil
import org.dbshell.db.metadata.DatabaseMetadata
import org.dbshell.environment.EnvironmentVars
import org.springframework.core.MethodParameter
import org.springframework.shell.CompletionContext
import org.springframework.shell.CompletionProposal
import org.springframework.shell.standard.ValueProvider
import org.springframework.stereotype.Component

@Component
class TableProvider: ValueProvider {
    override fun complete(
        completionContext: CompletionContext
    ): MutableList<CompletionProposal> {

        val currentInput = completionContext.currentWord()
        val types = arrayOf("TABLE")
        val dbmd = ConnectionInfoUtil.getConnectionFromCurrentContextJndi().connection.metaData
        val catalog = EnvironmentVars.currentCatalog
        val schema = EnvironmentVars.currentSchema

        return DatabaseMetadata.getTables(dbmd, catalog!!, schema!!, types)
                .map{ t -> t.tableName}
                .filter{ s -> s.contains(currentInput.toString())}
                .map{cp -> CompletionProposal(cp)}
                .toMutableList()
    }
}